src/app/Dashboard/DraggableRef.tsx

```diff
@@ -35,125 +35,327 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-import {
-  CardConfig,
-  dashboardCardConfigResizeCardIntent,
-} from '@app/Shared/Redux/Configurations/DashboardConfigSlicer';
-import { RootState } from '@app/Shared/Redux/ReduxStore';
-import { gridSpans } from '@patternfly/react-core';
-import _ from 'lodash';
+import { dashboardCardConfigReorderCardIntent } from '@app/Shared/Redux/Configurations/DashboardConfigSlicer';
+import { css } from '@patternfly/react-styles';
 import React from 'react';
-import { useDispatch, useSelector } from 'react-redux';
-import { DashboardCardSizes } from './Dashboard';
-import { DashboardCardContext } from './ResizableCard';
+import { useDispatch } from 'react-redux';
+
+type itemPosition = 'left' | 'right' | 'else';
+//
+// |-----------------| mouse |-----------------|
+// |   rect1.right   ^       ^   rect2.left    |
+// |                 |       |                 |
+// returns [inBetweenTwoRectangles, {insertedOnLeft, insertedOnRight, else}]
+function inBetween(ev: MouseEvent, rect1: DOMRect, rect2: DOMRect): [boolean, itemPosition] {
+  const withinHeightRect1 = ev.clientY > rect1.top && ev.clientY < rect1.bottom;
+  const withinHeightRect2 = ev.clientY > rect2.top && ev.clientY < rect2.bottom;
+  // Cases
+  // same row -> between: no ends
+  const singleRowBetween =
+    rect1.top === rect2.top &&
+    rect1.right <= ev.clientX + rect1.width * offsetScale &&
+    ev.clientX - rect2.width * offsetScale <= rect2.left &&
+    withinHeightRect1;
+  // same row -> before: left end
+  const singleRowBefore =
+    rect1.top === rect2.top && ev.clientX <= rect2.left && rect1.left >= rect2.right && withinHeightRect1;
+  // same row -> after: right end
+  const singleRowAfter =
+    rect1.top === rect2.top && ev.clientX >= rect1.right && rect1.right >= rect2.left && withinHeightRect1;
+
+  // different rows -> before: left end
+  const multRowBefore = rect1.top !== rect2.top && ev.clientX <= rect2.left && withinHeightRect2;
+  if (multRowBefore || singleRowBefore) {
+    return [true, 'left'];
+  }
+  // different rows -> after: right end
+  const multRowAfter = rect1.top !== rect2.top && ev.clientX >= rect1.right && withinHeightRect1;
+  if (multRowAfter || singleRowAfter) {
+    return [true, 'right'];
+  }
+  // different rows -> between: no ends
+  return [singleRowBetween, 'else'];
+}
+
+function getOrder(el: HTMLElement): number {
+  return parseInt(el.getAttribute(dashboardCardOrderAttribute) as string);
+}
+
+const initStyle = {};
+
+const transition = 'transform 0.5s cubic-bezier(0.2, 1, 0.1, 1) 0s';
```

Comment:
Maybe extra these (i.e. `transition` `delayedTransition`, `translateX` and `overlapTranslateY`)  to css rules. Then, we can add/remove from `Element.classList` as needed.

For  `translateX` and `overlapTranslateY`, we can specify a fix transform rule as we only have translation animation.

Reply:
Is toggling classes with specific css rules more performant than modifying the styles directly to the react ref? I can't find much about it.

Reply:
I would think the main performance cost of the animation is for the browser to actually do the animation, ie repaint frames to execute it. Which way the animation style is applied sounds like it shouldn't make much difference, so long as neither method is causing some additional re-renders or anything like that.

Putting the transition(s) into css and adding/removing from the class list might just be a little harder to muck up with typos, and it sounds like it's more suited to being easily extended with more potential animations in the future. To me that's the potential value I'm seeing from refactoring in this way.

Reply:
Agreed^^ also on this note, I found React docs say this:

https://reactjs.org/docs/faq-styling.html#are-inline-styles-bad